src/main/java/io/jenkins/plugins/coverage/CoveragePublisher.java

```diff
@@ -13,46 +14,55 @@
 import hudson.tasks.Recorder;
 import io.jenkins.plugins.coverage.adapter.CoverageReportAdapter;
 import io.jenkins.plugins.coverage.adapter.CoverageReportAdapterDescriptor;
-import io.jenkins.plugins.coverage.targets.CoverageMetric;
-import io.jenkins.plugins.coverage.targets.CoverageResult;
+import io.jenkins.plugins.coverage.exception.CoverageException;
 import io.jenkins.plugins.coverage.threshold.Threshold;
 import jenkins.tasks.SimpleBuildStep;
 import net.sf.json.JSONObject;
-import org.apache.commons.lang.StringUtils;
+import org.jenkinsci.Symbol;
 import org.kohsuke.stapler.DataBoundConstructor;
 import org.kohsuke.stapler.DataBoundSetter;
 import org.kohsuke.stapler.StaplerRequest;
 
 import javax.annotation.CheckForNull;
 import javax.annotation.Nonnull;
 import java.io.IOException;
+import java.util.LinkedList;
 import java.util.List;
 
 public class CoveragePublisher extends Recorder implements SimpleBuildStep {
 
-    private List<CoverageReportAdapter> adapters;
-    private List<Threshold> globalThresholds;
+    private List<CoverageReportAdapter> adapters = new LinkedList<>();
+    private List<Threshold> globalThresholds = new LinkedList<>();
+
+    private boolean failUnhealthy;
+    private boolean failUnstable;
+    private boolean failNoReports;
 
     private String autoDetectPath;
 
     @DataBoundConstructor
-    public CoveragePublisher(List<CoverageReportAdapter> adapters, List<Threshold> globalThresholds) {
-        this.adapters = adapters;
-        this.globalThresholds = globalThresholds;
+    public CoveragePublisher(String autoDetectPath) {
+        this.autoDetectPath = autoDetectPath;
     }
 
     @Override
     public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull TaskListener listener) throws InterruptedException, IOException {
-        CoverageProcessor processor = new CoverageProcessor(run, workspace, listener, adapters, globalThresholds);
+        listener.getLogger().println("Publishing Coverage report....");
 
-        if (!StringUtils.isEmpty(autoDetectPath)) {
-            processor.enableAutoDetect(autoDetectPath);
-        }
+        CoverageProcessor processor = new CoverageProcessor(run, workspace, listener);
+
+        processor.setAutoDetectPath(autoDetectPath);
 
-        CoverageResult result = processor.processCoverageReport();
+        processor.setFailUnhealthy(failUnhealthy);
+        processor.setFailUnstable(failUnstable);
+        processor.setFailNoReports(failNoReports);
 
-        CoverageAction action = new CoverageAction(result);
-        run.addAction(action);
+        try {
+            processor.performCoverageReport(getAdapters(), globalThresholds);
+        } catch (CoverageException e) {
+            listener.getLogger().println(e.getMessage());
```

Comment:
nit: It would be nice to log the full stack trace in a logger. The message for the user is good.